### PR TITLE
Convert inet_cache to a set

### PR DESCRIPTION
**Pull request by josevalim**
This patch is just a refactoring but it should provide
faster operations at the cost of slightly higher storage
(which is acceptable since the table is relatively small).
This will also make it possible to have decentralized
lookups in the future.

This is prep-work for the PR in #2818.

### CODE REVIEW

File: lib/kernel/src/inet_db.erl [hunk 1]
```diff
@@ -1621,46 +1618,39 @@ do_add_rr(RR, Db, State) ->
...
+    Time = times(),
+    case ets:lookup(inet_cache, Key) of
+	[{_,RR}] when RR#dns_rr.ttl =:= 0 -> %% at least once
+	    ets:delete(inet_cache, Key),
+	    [RR];
+	[{_,RR}] when RR#dns_rr.tm + RR#dns_rr.ttl < Time ->
+	    ets:delete(inet_cache, Key),
+	    [];
+	[{_,RR}] ->
+	    ets:update_element(inet_cache, Key, {2, RR#dns_rr{cnt = Time}}),
```
Comment (RaimoNiskanen) [inline lib/kernel/src/inet_db.erl:1646]: Since this function returns `true` or `false`, maybe indicate that: `_ = ets:update_element(...`

**Comment by RaimoNiskanen** [comment]
Looks fine to me.  I'll include it in our nightly tests.

File: lib/kernel/src/inet_db.erl (hunk 1 above)
Comment (josevalim) [reply lib/kernel/src/inet_db.erl]: Good call! I have added the underscore as well as a comment detailed why we are ignoring it.

File: lib/kernel/src/inet_db.erl [hunk 2]
```diff
@@ -1621,46 +1618,41 @@ do_add_rr(RR, Db, State) ->
...
+
+match_rr({_, _, _} = Key) ->
+    Time = times(),
+    case ets:lookup(inet_cache, Key) of
+	[{_,RR}] when RR#dns_rr.ttl =:= 0 -> %% at least once
+	    ets:delete(inet_cache, Key),
+	    [RR];
+	[{_,RR}] when RR#dns_rr.tm + RR#dns_rr.ttl < Time ->
+	    ets:delete(inet_cache, Key),
+	    [];
```
Comment (RaimoNiskanen) [inline lib/kernel/src/inet_db.erl:1644]: These lines are not covered in our daily tests by running all test suites in Kernel.

File: lib/kernel/src/inet_db.erl [hunk 3]
```diff
@@ -1746,16 +1738,17 @@ do_refresh_cache(CacheDb) ->
...
-    OldT = lists:foldl(Fun, OldestT, ets:lookup(CacheDb, Key)),
+    OldT =
+        case ets:lookup(CacheDb, Key) of
+	    [{_,RR}] when RR#dns_rr.tm + RR#dns_rr.ttl < Now ->
+		ets:delete(CacheDb, Key),
+		OldestT;
+	     [{_,#dns_rr{cnt = C}}] when C < OldestT ->
+		C;
+	     _ ->
+		OldestT
```
Comment (RaimoNiskanen) [inline lib/kernel/src/inet_db.erl:1750]: These lines are not covered in our daily tests by running all test suites in Kernel.

File: lib/kernel/src/inet_db.erl [hunk 4]
```diff
@@ -1800,17 +1793,18 @@ delete_older(CacheDb, TM, N) ->
...
-		  MM
+delete_older(Key, CacheDb, TM, N, M) ->
+    Next = ets:next(CacheDb, Key),
+    M1 =
+	case ets:lookup(CacheDb, Key) of
+	    [{_,RR}] when RR#dns_rr.cnt =< TM ->
+		ets:delete(CacheDb, Key),
+		M + 1;
+	    _ ->
+		M
```
Comment (RaimoNiskanen) [inline lib/kernel/src/inet_db.erl:1806]: These lines are not covered in our daily tests by running all test suites in Kernel.

File: lib/kernel/src/inet_db.erl [hunk 5]
```diff
@@ -1800,17 +1793,18 @@ delete_older(CacheDb, TM, N) ->
...
 	  end,
-    M1 = lists:foldl(Fun, M, ets:lookup(CacheDb, Domain)),
     delete_older(Next, CacheDb, TM, N, M1).
```
Comment (RaimoNiskanen) [inline lib/kernel/src/inet_db.erl:1808]: These lines are not covered in our daily tests by running all test suites in Kernel.

File: lib/kernel/src/inet_db.erl (hunk 2 above)
Comment (josevalim) [reply lib/kernel/src/inet_db.erl:1644]: Was this line covered in the previous implementation? I am asking just to know if it is something in my implementation or something that carried over from the previous code.

**Comment by RaimoNiskanen** [comment]
The PR looks fine in our daily tests.  I just went through the coverage logs after all Kernel suites, and that was a bit disappointing.

Nevertheless, having full coverage in daily tests of a PR is something I do not think we have as a blocking requirement.  I just wanted to complain a bit, not to you, but to whomever, that increased test coverage would be nice to have in addition to new features ;-)

File: lib/kernel/src/inet_db.erl (hunk 3 above)
Comment (josevalim) [reply lib/kernel/src/inet_db.erl:1750]: That makes sense. Everything is serialized on the server, there is no chance something may have been deleted meanwhile. Awesome catch, I will remove this and below.

**Comment by josevalim** [comment]
@RaimoNiskanen I agree! I can definitely help not make things worse and fix the coverage in the branches I added but I am afraid some of those are carried over from before my patch. I will look into them too but no guarantees. :)

**Comment by RaimoNiskanen** [comment]
@josevalim The test coverage problem is an old one; any help we might get would be much appreciated!

Thank you for your pull request!

File: lib/kernel/src/inet_db.erl (hunk 3 above)
Comment (RaimoNiskanen) [reply lib/kernel/src/inet_db.erl:1750]: Oh, sorry!  I thought it was a test case problem and merged.
Well, there will be future PR:s to do code cleanup in... ;-)

File: lib/kernel/src/inet_db.erl (hunk 2 above)
Comment (RaimoNiskanen) [reply lib/kernel/src/inet_db.erl:1644]: I have not checked, but would not be surprised if they were not.  Our tests pay too little attention to `inet_res`.

File: lib/kernel/src/inet_db.erl (hunk 3 above)
Comment (josevalim) [reply lib/kernel/src/inet_db.erl:1750]: I will clean it up on the PR that removes the deletes and makes it concurrent. :)

Comment (josevalim) [reply lib/kernel/src/inet_db.erl:1750]: Good news, those are needed for the PR that makes it concurrent. :) I am submitting it soon!

**Pull request merged**